substrate/client/rpc-spec-v2/src/archive/archive_storage.rs

```diff
@@ -127,3 +139,724 @@ where
 		ArchiveStorageResult::ok(storage_results, discarded_items)
 	}
 }
+
+/// Parse hex-encoded string parameter as raw bytes.
+///
+/// If the parsing fails, returns an error propagated to the RPC method.
+pub fn parse_hex_param(param: String) -> Result<Vec<u8>, ArchiveError> {
+	// Methods can accept empty parameters.
+	if param.is_empty() {
+		return Ok(Default::default())
+	}
+
+	array_bytes::hex2bytes(&param).map_err(|_| ArchiveError::InvalidParam(param))
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct DiffDetails {
+	key: StorageKey,
+	return_type: ArchiveStorageDiffType,
+	child_trie_key: Option<ChildInfo>,
+	child_trie_key_string: Option<String>,
+}
+
+/// The type of storage query.
+#[derive(Debug, PartialEq, Clone, Copy)]
+enum FetchStorageType {
+	/// Only fetch the value.
+	Value,
+	/// Only fetch the hash.
+	Hash,
+	/// Fetch both the value and the hash.
+	Both,
+}
+
+pub struct ArchiveStorageDiff<Client, Block, BE> {
+	client: Storage<Client, Block, BE>,
+}
+
+impl<Client, Block, BE> ArchiveStorageDiff<Client, Block, BE> {
+	pub fn new(client: Arc<Client>) -> Self {
+		Self { client: Storage::new(client) }
+	}
+}
+
+impl<Client, Block, BE> ArchiveStorageDiff<Client, Block, BE>
+where
+	Block: BlockT + 'static,
+	BE: Backend<Block> + 'static,
+	Client: StorageProvider<Block, BE> + Send + Sync + 'static,
+{
+	/// Fetch the storage from the given key.
+	///
+	/// This method returns:
+	/// - `None` if the storage is not present.
+	/// - `Some((StorageResult, None))` for `FetchStorageType::Value`.
```

Comment:
would be nice to provide an enum for the different return types such as:

```rust
enum FetchStorageType {
	/// Only fetch the value.
	Value(StorageResult)
	/// Only fetch the hash.
	Hash(StorageResult)
	/// Fetch both the value and the hash.
	Both { value: StorageResult. hash: StorageResult }
}
```

Reply:
Looks good! Thanks for the suggestion